larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+    {
+        hitCounter++;
+        const float pCaloHitCharge(pCaloHit->GetInputEnergy());
+
+        if (pCaloHitCharge < 0)
+        {
+            std::cout << "Found a hit with negative charge! " << std::endl;
+        }
+        else
+        {
+            totalCharge    += pCaloHitCharge;
+            chargeVector.push_back(pCaloHitCharge);
+
+            if (hitCounter <= std::ceil(totalHits*endChargeFraction)) //ensure at least 1 hit is used
+            {
+                startCharge += pCaloHitCharge;
+            }
+            if (hitCounter >= std::floor(totalHits*(1.f-endChargeFraction)))
+            {
+                endCharge += pCaloHitCharge;
+            }
+
+            if(hitCounter>1)
+                incrementCharge = (pCaloHitCharge-previousCharge)*(pCaloHitCharge-previousCharge);
+
+            previousCharge = pCaloHitCharge;
+        }
+    }
+
+    chargeMean = 0.f;
+    chargeSigma = 0.f;
+
+    for (const float charge : chargeVector)
+        chargeMean += charge;
+
+    chargeMean /= static_cast<float>(chargeVector.size());
+
+    for (const float charge : chargeVector)
+        chargeSigma += (charge - chargeMean) * (charge - chargeMean);
+
+    if (chargeSigma < 0.f)
+        throw StatusCodeException(STATUS_CODE_FAILURE);
+
+    chargeSigma = std::sqrt(chargeSigma / static_cast<float>(chargeVector.size()));
+    incrementCharge = std::sqrt(incrementCharge / static_cast<float>(chargeVector.size()));
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+CaloHitList ChargeFeatureTool::OrderCaloHitsByDistanceToVertex(const Algorithm *const pAlgorithm, const pandora::Cluster *const pCluster)
+{
+    //find the neutrino vertex and sort hits by distance to vertex
+    const VertexList *pVertexList = nullptr;
+    (void) PandoraContentApi::GetCurrentList(*pAlgorithm, pVertexList);
+
+    if (!pVertexList || (pVertexList->size() != 1) || (VERTEX_3D != pVertexList->front()->GetVertexType()))
+        throw StatusCodeException(STATUS_CODE_FAILURE);
+
+    const Vertex *const pVertex(pVertexList->front());
+    const HitType hitType(LArClusterHelper::GetClusterHitType(pCluster));
+
+    const CartesianVector vertexPosition2D(LArGeometryHelper::ProjectPosition(pAlgorithm->GetPandora(), pVertex->GetPosition(), hitType));
+
+    CaloHitList clusterCaloHitList;
+    pCluster->GetOrderedCaloHitList().FillCaloHitList(clusterCaloHitList);
+    CaloHitVector clusterCaloHitVector(clusterCaloHitList.begin(), clusterCaloHitList.end());
+
+    //TODO: might give problems if vertex in the middle of the cluster ?
+    std::sort(clusterCaloHitVector.begin(), clusterCaloHitVector.end(), std::bind(SortByDistanceToVertex, std::placeholders::_1, std::placeholders::_2, vertexPosition2D));
+    CaloHitList orderedCaloHitList(clusterCaloHitVector.begin(),clusterCaloHitVector.end());
+
+    return orderedCaloHitList;
+}
```

Comment:
Maybe provide orderedCaloHitList as a non-const reference argument; can then fill and sort the container directly, rather than returning a copy.

Reply:
ok